frontends/pytorch/csrc/builder/graph_importer.h

```diff
@@ -0,0 +1,117 @@
+//===- graph_importer.h -----------------------------------------*- C++ -*-===//
+//
+// This file is licensed under a pytorch-style license
+// See frontends/pytorch/LICENSE for license information.
+//
+//===----------------------------------------------------------------------===//
+
+#ifndef NPCOMP_FRONTENDS_PYTORCH_CSRC_GRAPH_IMPORTER_H
+#define NPCOMP_FRONTENDS_PYTORCH_CSRC_GRAPH_IMPORTER_H
+
+#include <memory>
+
+#include "../pybind.h"
+#include "func_builder.h"
+
+#include "mlir-c/IR.h"
+
+#include <torch/csrc/jit/api/compilation_unit.h>
+#include <torch/csrc/jit/ir/ir.h>
+
+namespace torch_mlir {
+
+/// Main entry-point for importing graphs.
+/// In torch terminology, a Graph is a function. We materialize it as at
+/// least one MLIR function, but may specialize multiple versions of it,
+/// depending on usage.
+///
+/// Since graph functions typically have enough annotations for the most
+/// generic form of every type (i.e. Tensor, List, etc), and since we often
+/// want to multi-version specific specializations, we take the approach of
+/// generating a '$generic' suffixed function at that level and then generate
+/// the actual named function with using a 'numpy.generic_call' op to invoke
+/// the generic function with metadata controlling how it is legal to
+/// specialize. This leaves the process of inlining and expanding the
+/// specializations to compiler passes.
+class GraphImporter : public std::enable_shared_from_this<GraphImporter> {
+public:
+  struct MlirMappingOptions {
+    MlirContext context;
+    std::string genericFuncName;
+    std::string funcName;
+    TypeMapper &typeMapper;
+    FuncBuilder::Inserter &inserter;
+  };
+  /// Construct an importer.
+  GraphImporter(std::shared_ptr<torch::jit::Graph> graph,
+                MlirMappingOptions mappingOptions);
+
+  /// Helper to create a graph importer from a traced/scripted python function.
+  /// If the funcName of the mapping options is not set, it is set from the
+  /// function name. It is the responsibility of the caller to ensure that the
+  /// funcObj and associated graph outlives this instance.
+  static std::shared_ptr<GraphImporter>
+  forPythonJitFunc(torch::jit::Function *function,
+                   MlirMappingOptions mappingOptions);
+
+  /// Initialize for import. This is separate from the constructor purely for
+  /// ergonomics and must be called post-construction. Initialization activities
+  /// that throw go here.
+  void initialize();
+
+  /// Imports the generic function into the module.
+  void importGenericFunc();
+
+private:
+  // A scope of Graph Value * to corresponding MlirValue. Scopes nest
```

Comment:
Why do we need scopes? Aren't the Value*'s unique in the Graph?

Reply:
Yes, but just looking ahead a bit to dealing with nested "blocks" (MLIR Regions) for If/While, etc. They implicitly capture but contained values can't be accessed outside of the block. I was going to handle the nesting with a scope, which might help me with some of the boundary mappings.